### ISSUE TITLE
Indentation: JSX child element multiline prop gets unexpected extra indentation level

### ISSUE THREAD

**Issue by unframework** [eslint/eslint#10884]
**Tell us about your environment**

* **ESLint Version:** 5.6.0
* **Node Version:** 9.8.0
* **npm Version:** 5.6.0

**What parser (default, Babel-ESLint, etc.) are you using?** Babel-ESLint

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
    "env": {
        "browser": true,
        "commonjs": true,
        "es6": true
    },
    "extends": "eslint:recommended",
    "parser": "babel-eslint",
    "parserOptions": {
        "ecmaFeatures": {
            "jsx": true
        },
        "sourceType": "module"
    },
    "plugins": [
        "babel",
        "react"
    ],
    "rules": {
        "indent": [ "error", 4 ],
        "indent-legacy": "off",
        "jsx-quotes": [ "error", "prefer-double" ]
    }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```jsx
export default <test prop={
    null
}><inner prop={
    null
} /></test>;
```

```bash
node_modules/.bin/eslint test.jsx
```

**What did you expect to happen?** No errors

**What actually happened? Please include the actual, raw output from ESLint.**

```
$ node_modules/.bin/eslint test.jsx

/path/to/project/test.jsx
  4:1  error  Expected indentation of 8 spaces but found 4  indent
  5:1  error  Expected indentation of 4 spaces but found 0  indent

✖ 2 problems (2 errors, 0 warnings)
  2 errors and 0 warnings potentially fixable with the `--fix` option.
```

---

My initial encounter with this behaviour was with more complex use cases (JSX function-as-a-child pattern) but the above sample syntax is the simplest one that still triggers the indentation error.

In comparison, the following JSX syntax does *not* trigger the indentation error:

```jsx
export default <test prop={null}><inner prop={
    null
} /></test>;
```

Looking at `indent.js`, I think the place that effects this result is the JSX element handler:

```js
// ...
            JSXElement(node) {
                if (node.closingElement) {
                    addElementListIndent(node.children, sourceCode.getFirstToken(node.openingElement), sourceCode.getFirstToken(node.closingElement), 1);
                }
            },
// ...
```

The children of an element are expected to be 1 level in, relative to the starting token of the opening tag - which is the `<`. However, if the child starts on the same line as the opening tag, that offset is nullified (per logic in `getDesiredIndent(token)` function).

This is why the second JSX example does not trigger the indentation error - the inner element starts on same line as the starting `<` of the outer element. But in the first JSX example the inner element is no longer on the same line as the outer element's `<` token, and that makes the lint rule expect one more indentation level for the inner element's prop value.

The obvious workaround is to just put `<inner ... />` on a new line and indent it. But it feels like using the starting `<` as indentation offset reference is a bit too simplistic? Just doesn't feel completely "right"; also, we use a lot of function-as-a-child syntax and it would help save on indentation levels.

In my tests, I changed the JSX element handler logic to use the last token of the opening tag (i.e. the `>` that immediately precedes the contents), and that worked really well. If the child literal or expression starts on a new line, then the indentation level is increased by one as intended, but if it is on the same line, then the offset gets correctly ignored and the inner multiline props do not trigger the indentation error.

However, there is a problem with that approach - in some style guides, the opening tag's `>` is expected to be indented relative to the `<`: i.e the `react/jsx-closing-bracket-location` rule with `props-aligned` setting. Thus, the following code would be valid but still trip up `indent` if the `>` is used as offset reference:

```
export default <test
    prop={null}
    >
    Some content
</test>;
```

I am not sure how many folks are using that convention - the `indent` rule actually "fights" it anyway as is - but that is still a caveat.

Thoughts? If there is a configuration option that might apply, happy to try it.

Thanks!

**Comment by ilyavolodin** [comment]
@not-an-aardvark Is this actually a bug?

**Comment by unframework** [comment]
Thanks for the response! I know that that question wasn't aimed at me, but just wanted to clarify that to me it felt unexpected that this works:

```jsx
export default <test prop={null}><inner prop={
    null
} /></test>;
```

... and this doesn't, after adding newlines to the outer element's prop:

```jsx
export default <test prop={
    null
}><inner prop={
    null
} /></test>;
```

The error does not even happen on `<inner ...` but on the second `null`. Bug or not, this behaviour just felt hard to understand/explain without digging into the lint rule source code.

Thanks again!

**Comment by ilyavolodin** [comment]
The reason why I'm asking @not-an-aardvark is that he is the creator of the current version of the rule. And to me personally, both of the examples look incorrect. I would expect the correct code to look like this:
```js
export default <test prop={null}><inner prop={
        null
    } /></test>;
```
Since one indent level is added by opening JSX tag, and another - by multi-line `export` statement.

**Comment by unframework** [comment]
Oh, yeah I didn't mean to confuse things with the `export`, I should have used a simple return statement as an example probably...

**Comment by not-an-aardvark** [comment]
This seems like a bug. The `/>` in a self-closing JSX element should be at the same indentation level as the first token of the element. The [current implementation](https://github.com/eslint/eslint/blob/066f7e07d4f722573c6dbe7f64b0765504595644/lib/rules/indent.js#L1402-L1406) doesn't apply any offset to self-closing elements, which causes problems when a self-closing element starts in the middle of a line ([simplified example](https://eslint.org/demo/#eyJ0ZXh0IjoiLyogZXNsaW50IGluZGVudDogZXJyb3IgKi9cclxuPHRlc3RcclxuPjxpbm5lciBcclxuLz48L3Rlc3Q+OyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6OCwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6eyJqc3giOnRydWV9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)).

**Comment by unframework** [comment]
Would a possible fix be changing [this line](https://github.com/eslint/eslint/blob/066f7e07d4f722573c6dbe7f64b0765504595644/lib/rules/indent.js#L1404) from:

```js
addElementListIndent(node.children, sourceCode.getFirstToken(node.openingElement), sourceCode.getFirstToken(node.closingElement), 1);
```

to:

```js
addElementListIndent(node.children, sourceCode.getLastToken(node.openingElement), sourceCode.getFirstToken(node.closingElement), 1);
```

(instead of `getFirstToken` using `getLastToken` for the opening element)?

This worked pretty well in my local testing, but I am sure there are a lot of other implications to consider... Either way, happy to pitch in a PR if there is consensus on a fix.

**Comment by unframework** [comment]
Oops, looks like this sat too long on my to-do pile. Would it be possible to reopen this issue? It has been accepted as worth fixing at least, and I'd still like to have a crack at this.

Also, [here's a live demo in the online sandbox](https://eslint.org/demo/#eyJ0ZXh0IjoiY29uc3QgYSA9IDx0ZXN0IHByb3A9e1xuICAgIG51bGxcbn0+PGlubmVyIHByb3A9e1xuICAgIG51bGxcbn0gLz48L3Rlc3Q+O1xuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo5LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7ImpzeCI6dHJ1ZX19LCJydWxlcyI6eyJpbmRlbnQiOjJ9LCJlbnYiOnsibm9kZSI6dHJ1ZSwicGhhbnRvbWpzIjpmYWxzZX19fQ==) - it is the same example as above, just adding here for extra reference.

Thanks!

**Comment by platinumazure** [comment]
Yes, we can reopen this, no worries. Thanks for your interest!

**Comment by unframework** [comment]
Folks, this turned out to be not enough of a priority to put out a polished-enough fix on my end - I just got used to a more indent-heavy code style anyway. Thanks for helping discuss this!
